include/operations/Operation.hpp

```diff
@@ -88,7 +88,20 @@ class Operation {
 
   virtual void setTargets(const Targets& t) { targets = t; }
 
-  virtual void setControls(const Controls& c) { controls = c; }
+  virtual void setControls(const Controls& c) {
+    clearControls();
+    addControls(c);
+  }
+
+  virtual void addControl(const Control c) { addControls({c}); }
+
+  virtual void addControls(const Controls& c) = 0;
```

Comment:
It's a minor detail, but I would have rather expected the `addControl` method to be abstract and the `addControls` method to just be
```c++
for (const auto& control: c) {
  addControl(control);
}
```
Kind of feels more natural to be to describe how a single control is added and then just reusing that method for the multiple control case. That should also safe some code duplication of the loops over controls.